### ISSUE TITLE
parsing of timeout option fails - order of nameservers used is reversed

### ISSUE THREAD

**Issue by birnbacs** [mikepultz/netdns2#34]
My resolv.conf contains settings for a first and a second nameserver:
nameserver 192.168.0.10
nameserver 192.168.0.1

The latter (~.1) is also the GW for my subnet so it is used only as a backup if the dedicated nameserver (~.10) fails. 

When using Horde, for some reason a DNS query is processed on each and every HTTP request. My problem was that every request took 5 seconds to process while waiting for -- what? 

On this, DNS2 throws errors:

Jun 14 21:35:00 max HORDE: PHP ERROR: Undefined offset: 1 [pid 6887 on line 416 of "/usr/local/share/pear/Net/DNS2.php"]

I experimentally put some debugging code two lines above the referenced statement to dump $line and this it what it yields:

...
Jun 14 21:41:51 max HORDE: line is <my.domain>
Jun 14 21:44:55 max HORDE: line is nameserver 192.168.0.10
Jun 14 21:44:55 max HORDE: line is options:timeout:1
Jun 14 21:44:55 max HORDE: PHP ERROR: Undefined offset: 1 [pid 6934 on line 418 of "/usr/local/share/pear/Net/DNS2.php"]
...

Obviously the timeout options for DNS lookup doesn't get processed properly. It also seems that the secondary server (~.10) is queried first and the primary (~.1) only after the secondary server failed. 

Now, in my environment the secondary had died so that a request to it would take until the timeout. The default timeout is 5 seconds, just the delay I was looking for. After the timeout the primary DNS was queried and everything functioned again. 

Curing this was easy enough for me, I just commented out the secondary nameserver in /etc/resolv.conf. Still, the two issues above should get a good looking-at.

**Comment by mikepultz** [comment]
Thanks for your bug report.

Can you let me know what version of Net_DNS2 that is? The version is listed in the Net/DNS2.php file near the top of the class definition (const VERSION = ).

Also- can you send me the full contents of your resolv.conf file? Just so I can re-create it exactly on my side, and see if it's a parsing issue or something else.

Thanks!

Mike

**Comment by birnbacs** [comment]
Mike,

const VERSION = '1.4.1';

The resolv.conf reads:

[file begin]
domain mueller-schubert.com
nameserver 192.168.0.10
nameserver 192.168.0.1
options:timeout:1
[file end]

I see I have a syntax error in the options line, the first colon should be
a blank. Parsing this line by DNS2 is probably OK.


Consider this:

[develop@max /usr/local/www/apache24/data]$ nslookup heise.de 192.168.0.10
Server:         192.168.0.10
Address:        192.168.0.10#53

Non-authoritative answer:
Name:   heise.de
Address: 193.99.144.80

[develop@max /usr/local/www/apache24/data]$ nslookup heise.de 192.168.0.1
;; connection timed out; no servers could be reached

[develop@max /usr/local/www/apache24/data]$ nslookup heise.de
Server:         192.168.0.10
Address:        192.168.0.10#53

Non-authoritative answer:
Name:   heise.de
Address: 193.99.144.80

[develop@max /usr/local/www/apache24/data]$



If there is anything else I can do to help you fix the problem I will be
glad to help.

Regards

  Sebastian



2015-06-17 16:36 GMT+02:00 Mike Pultz <notifications@github.com>:

**Comment by mikepultz** [comment]
Hey Sebastian,

So the syntax error was just the missing space in the options line- when I was pulling the data, I was splitting the values by space, without checking them first- so I'm checking that now- that error will go away.

The timeout- like you say, the default is set to 5 seconds in the library- and right now I'm not parsing the options setting in resolv.conf to set a different timeout- so it makes sense why you're seeing that.

What I can't reproduce, is it using the second name server before the first- in all my testing, it uses the name servers in order, unless the "ns_random" flag was set to true in Net_DNS2- in which case, it randomizes the list before using it.

Can you check to see- when Net_DNS2 is initialized in Horde, is it setting the ns_random flag?

Mike

**Comment by birnbacs** [comment]
Dear Mike,

thanks for investigating this.

2015-06-22 7:15 GMT+02:00 Mike Pultz <notifications@github.com>:

> Hey Sebastian,
>
> So the syntax error was just the missing space in the options line- when I
> was pulling the data, I was splitting the values by space, without checking
> them first- so I'm checking that now- that error will go away.
>
Great.

> The timeout- like you say, the default is set to 5 seconds in the library-
> and right now I'm not parsing the options setting in resolv.conf to set a
> different timeout- so it makes sense why you're seeing that.
>
As I said, there was a syntax error on my side, too, there should have been
at least one whitespace in the line.  After fixing this I enjoy the reduced
timeout by the regular mechanism.

> What I can't reproduce, is it using the second name server before the
> first- in all my testing, it uses the name servers in order, unless the
> "ns_random" flag was set to true in Net_DNS2- in which case, it randomizes
> the list before using it.
>
> Can you check to see- when Net_DNS2 is initialized in Horde, is it setting
> the ns_random flag?
>

I'll be glad to - only how do I go about it?
I have no idea how to check or set such flags.

Kind regards

  Sebastian

**Comment by mikepultz** [comment]
Hey Sebastian,

Good question- so somewhere in the horde code it's going to create a new instance of the Net_DNS2 object- so something like "$x = new Net_DNS2_Resolver"- options are passed either directly in the constructor as an array, or set in the object by setting "$x->ns_random = true".

if you're not sure how to find it, just let me know what version of horde it is, and I can see if I can take a look.

Mike

**Comment by birnbacs** [comment]
Hi Michael,

my Horde is version 5.2.5.

I grepped for any occurrence of 'random' in all of Horde's *.php files but
could not get a hit:

[develop@max /usr/local/www/apache24/data]$ sudo grep random `find . -type
f -name "*.php"`
[develop@max /usr/local/www/apache24/data]$

Net_DNS2_Resolver is instantiated twice inside Horde:

[develop@max /usr/local/www/apache24/data]$ sudo grep Net_DNS2_Resolver
`find . -type f -name "*.php"`
./lib/LoginTasks/Task/LastLogin.php:        if ($dns =
$injector->getInstance('Net_DNS2_Resolver')) {
./admin/sessions.php:    $resolver =
$injector->getInstance('Net_DNS2_Resolver');

I could not find any suspect expression that hinted at the random variable,
though.

If you tell me where I shall introduce a debugging message in the DNS2-code
so we know for sure whether or not random is active.


Kind regards

  Sebastian


2015-06-23 5:01 GMT+02:00 Mike Pultz <notifications@github.com>:

**Comment by mikepultz** [comment]
Hey Sebastian,

in DNS2.php, around line 885, it says:

if ($this->ns_random == true) {

if you echo out the value of ns_random just before this, we can see what it's set to- so something like:

echo ($this->ns_random == true) ? "enabled" : "disabled";

Let me know!

Mike

**Comment by birnbacs** [comment]
Dear Mike,

I checked an the answer is yes. And no, too.



First I logged with this line in DND2.php (line 872, right before the
conditional shuffle):

syslog(LOG_ERR, "ns_random is " . ($this->ns_random == true) ? "enabled" :
"disabled");

which produced this log output:

Jun 25 13:23:17 max HORDE: enabled
Jun 25 13:23:38 max last message repeated 4 times



On the other hand, this debug statement:

syslog(LOG_ERR, "DNS2: ns_random is " . $ns_random . ".");

produced:

Jun 25 13:20:16 max HORDE: DNS2: ns_random is .



It appears that the variable is not set at all. Does that make sense?


Funny, I had a similar error on my Mac this morning. The second-line
nameserver got queried first and the first-line nameserver second.
Both the Mac and my Horde installation utilize BSD - in different flavours
and on different networks, though.

From the local manpage for resolv.conf everything looks straightforward:

     nameserver  Internet address (in dot notation) of a name server that
the
                 resolver should query.  Up to MAXNS (currently 3) name
                 servers may be listed, one per keyword.  If there are
multi-
                 ple servers, the resolver library queries them in the order
                 listed.


Kind regards

  Sebastian


2015-06-24 23:49 GMT+02:00 Mike Pultz <notifications@github.com>:

**Comment by mikepultz** [comment]
Hey Sebastian,

Yeah, the test is a little unclear when you pass it to syslog()- to make it work right with other text, you need to wrap it in another set of braces (if you notice, it just said "enabled" not "ns_random is enabled".

So I think this should give you the real answer:

syslog(LOG_ERR, "ns_random is " . (($this->ns_random == true) ? "enabled" : "disabled"));

Your second test, you also just used "$ns_random" and not "$this->ns_random"- just "$ns_random" isn't actually set, so it's not weird that it was empty.

Let me know what the test says,

Mike

**Comment by birnbacs** [comment]
Mike,

sorry for the incorrect logging. Now I get this:

Jun 26 10:21:55 max HORDE: [horde] Login success for sbirnbach to horde
(192.168.0.106) [pid 87890 on line 164 of
"/usr/local/www/apache24/data/login.php"]
Jun 26 10:21:55 max HORDE: ns_random is disabled

I am open to carry out any other test you may require.

BTW: forget about the Mac issue, it has nothing to do with BSD as OSX
employs a different name resolution system that is just mapped to
resolv.conf

Regards

  Sebastian

2015-06-26 1:22 GMT+02:00 Mike Pultz <notifications@github.com>:

**Comment by mikepultz** [comment]
Hey Sebastian,

Sorry about the late reply- just been busy with other stuff-

Can you try changing that line to:

syslog(LOG_ERR, "servers are: " . print_r($this->nameservers, true));

and see what it returns?

your debug shows that the randomize flag isn't set- which is about the only thing I can think of that would make it use the servers out of order- this will show the contents of the servers array and see what order they're in.

Thanks!

Mike

**Comment by birnbacs** [comment]
Hey Mike,

I introduced said statement right before the conditional shuffling (line
868 of DNS2.php) and this is what I get:

Jul  6 13:00:49 max HORDE: servers are: Array (     [0] => 192.168.0.10
[1] => 192.168.0.1 )
Jul  6 13:00:54 max HORDE: [horde] Login success for sbirnbach to horde
(192.168.0.106) [pid 17724 on line 164 of
"/usr/local/www/apache24/data/login.php"]
Jul  6 13:00:54 max HORDE: servers are: Array (     [0] => 192.168.0.10
[1] => 192.168.0.1 )


I hope that gives you a clue. If not please don't hesitate to write back
for more tests.

Kind regards

  Sebastian

2015-07-06 1:35 GMT+02:00 Mike Pultz <notifications@github.com>:

**Comment by bishopb** [comment]
Based on the conversation and the provided debugging statements, it appears that:

* the timeout issue has been resolved (`timeout:1` now correctly seen)
* the nameservers are listed in the correct order
* `ns_random` is false

@birnbacs, are you still experiencing a delay?  If so, please re-run with the debugging suggested above and paste the results (just to make sure we're all on the same page).  Also, I noticed that Horde creates a [resolver object with a cache][1]. Try deleting that cache (possibly at `/tmp/horde_dns.cache`) and see if that helps resolve the issue.

[1]: https://github.com/horde/horde/blob/0e8ec162572d8e21a78c75e550512ada0e2fa1af/framework/Core/lib/Horde/Core/Factory/Dns.php#L16

**Comment by birnbacs** [comment]
Dear Mike,

thanks for digging so deep into this. It seems the problems I experienced
were caused by a mixture of a badly administered DNS, an incorrect
resolv.conf entry and possibly a parsing error. After the latest changes
and tests I have not been experiencing any more DNS delays, as far as I am
concerned the bug seems fixed.

Kind regards

  Sebastian

2016-01-05 16:08 GMT+01:00 Bishop Bettini <notifications@github.com>:
